### PR TITLE
Thoroughly add “bucket type” argument

### PR DESCRIPTION
**Pull request by llelf**

### CODE REVIEW

**Comment by tmcgilchrist** [comment]
Thanks, just need some time to read through bucket types and review this code. 

Ping us if I haven't got back to this in a few days.

File: src/Network/Riak/Basic.hs [hunk 1]
```diff
@@ -77,40 +77,40 @@ getServerInfo conn = exchange conn Req.getServerInfo
...
 -- responsibility.
 --
 -- You should /only/ supply 'Nothing' as a 'T.VClock' if you are sure
--- that the given bucket+key combination does not already exist.  If
--- you omit a 'T.VClock' but the bucket+key /does/ exist, your value
--- will not be stored.
-put :: Connection -> T.Bucket -> T.Key -> Maybe T.VClock
+-- that the given type+bucket+key combination does not already exist.
+-- If you omit a 'T.VClock' but the type+bucket+key /does/ exist, your
+-- value will not be stored.
```
Comment (tmcgilchrist) [inline src/Network/Riak/Basic.hs]: This behaviour of not storing the value is concerning, is there any indication in the response about this failure? Could the types be modified to better express this?

Comment (tmcgilchrist) [inline src/Network/Riak/Basic.hs]: Actually that was already an issue, unrelated to these changes.

I'm happy with what's here, thanks @llelf

**Comment by sanketr** [comment]
When will this change be available in a release on stackage? 

I am using v1.0.0.1 from stackage which doesn't seem to have "Bucket Type" argument under `Network.Riak.Basic`. Since in Riak 2.0, bucket types are additional namespaces, this is very useful thing to have. In my case, I use bucket type every where as namespace to separate out different collections. So, without bucket type, I can't make a get/put call.

**Pull request merged**